### DISCUSSION TITLE
translate-x-0 and -translate-x-0 are written before translate-x-full, thus creating a right sidebar is impossible

### DISCUSSION THREAD

**Discussion by GeorgeFlorian** [tailwindlabs/tailwindcss#12248] [Help] [answered]
**What version of Tailwind CSS are you using?**

"tailwindcss": "^3.3.3"

**What version of Node.js are you using?**

node -v
v18.15.0

**What browser are you using?**

Chrome, Firefox

**What operating system are you using?**

Ubuntu 23

**Reproduction URL**

I am not able to create a proper example, because I can't make alpineJS in CodePen. The following code works in my project:
https://codepen.io/bleah1/pen/eYxOpOM

**Describe your issue**

I have two sidebars, one on the left and one on the right. Each of them is opened by identical buttons (simplified for this example):
```html
  <button type="button"
          class="inline-flex items-center p-2 text-sm text-black rounded-lg lg:hidden bg-white focus:outline-none focus:ring-2 focus:ring-gray-200 hover:text-gray-400"
          @click="open = true">
    <span class="bg-gray-500">Click to Open</span>
  </button>
```

The one on the left works as intended, but the one on the right does not.

The one on the left:
```html
  <aside class="fixed top-0 left-0 z-40 w-64 h-screen transition-transform -translate-x-full lg:translate-x-0"
         :class="{ 'translate-x-0': open }"
         aria-label="Sidebar">
```

The one on the right:
```html
  <aside class="fixed top-0 right-0 z-40 w-64 h-screen transition-transform translate-x-full"
         :class="{ 'translate-x-0': open }"
         aria-label="Profile Sidebar">
  
```
Before clicking the open button:
### Left sidebar:
![image](https://github.com/tailwindlabs/tailwindcss/assets/39895513/dc930528-7d14-40ba-991f-3b6cffb5cc80)
### Right sidebar:
![image](https://github.com/tailwindlabs/tailwindcss/assets/39895513/85f6d15c-2a90-49b4-a945-ab19bd16909a)

After clicking the open button, this is the CSS that's being applied to the `<aside>` tags:
### The one on the left:
![image](https://github.com/tailwindlabs/tailwindcss/assets/39895513/29e5c616-0beb-4ac7-a8a6-d3f8a95a14ce)

### The one on the right:
- using `translate-x-0`:
![image](https://github.com/tailwindlabs/tailwindcss/assets/39895513/7de3db80-daf8-474a-b0d7-da18d09591e8)
- using `-translate-x-0`:
![image](https://github.com/tailwindlabs/tailwindcss/assets/39895513/fd7a1051-2136-42d2-b38a-dd940adc954a)

`-translate-x-full` is at `output.css:995`
`translate-x-full` is at `output.css:1010`
`translate-x-0` is at `output.css:1005`
`-translate-x-0` is at `output.css:990`

### As you can see, no matter what `translate-x-0` I use, they both are placed before `translate-x-full` in `output.css`.
So, it's impossible to create a sidebar from the right using this method.

**Is it a bug or the intended behavior ?
How can I create a sidebar that's coming from the right ?**

I have also thought of writing the classes like this:
```html
<aside 
    class="fixed top-0 right-0 z-40 w-64 h-screen transition-transform" 
    :class="open ? 'translate-x-0' : 'translate-x-full'" 
    aria-label="Sidebar">
    <!-- Sidebar content goes here -->
</aside>
```
But this creates a flashing of the sidebar when loading any page.

**Comment by MartijnCuppens** [comment]
> Is it a bug or the intended behaviour ?

You're not supposed to add utilities that will control the same property, that's why it behaves so weird 😀

> I have also thought of writing the classes like this:
> ...
> But this creates a flashing of the sidebar when loading any page.

This probably means the value of `open` changes during loading, maybe avoiding that will help? 

An other solution which will help if is to make use of the [important modifier](https://tailwindcss.com/docs/configuration#important-modifier) to prioritise certain classes?

**Comment by alex-krasikau** [comment]
Hey @GeorgeFlorian!

@MartijnCuppens is correct, you shouldn't rely on the class declaration order in the resulting CSS. You can make your code resistant to this issue if you only render one of the conflicting classes at a time.

So this is not the bug, but the expected behavior of the CSS (not Tailwind).

The last example you provided is the correct solution: `:class="open ? 'translate-x-0' : 'translate-x-full'"`. But in order to fix the flash on load, you might need to use something less elegant, like the important modifier: `:class="{ '!translate-x-0': open }"`, hide the sidebar until Alpine picks up your element: `class="hidden" :class="{ 'block': open }"`. Or figure out something similar.

**Reply by GeorgeFlorian** [reply]
@alex-krasikau 
I am struggling figuring out a fix for this.
One of the sidebar must be displayed on large devices: `lg:translate-x-0`
```html
<aside class="fixed top-0 left-0 z-40 w-64 h-screen transition-transform -translate-x-full lg:translate-x-0"
         :class="{ 'translate-x-0': open }"
         aria-label="Sidebar">
```

How can I control 3 cases (for mobile opened and closed, for desktop opened) with just `:class` and `open` ?

**Reply by alex-krasikau** [reply]
Looks like adding an important modifier to the `translate-x-0` should do the trick:
```
<aside class="fixed top-0 left-0 z-40 w-64 h-screen transition-transform -translate-x-full lg:translate-x-0"
         :class="{ '!translate-x-0': open }"
         aria-label="Sidebar">
```

Let me know if you still experience issues.

**Reply by GeorgeFlorian** [reply]
I've added the utility classes conditionally and I've used Alpine's `x-cloak` [directive](https://alpinejs.dev/directives/cloak) to stop the flashing:
```html
  <aside x-cloak class="fixed top-0 left-0 z-40 w-64 h-screen transition-transform"
         :class="{ '-translate-x-full': !open, 'translate-x-0': open, 'lg:translate-x-0': true}"
         aria-label="Sidebar">
```

**Reply by alex-krasikau** [reply, marked as answer]
Yeah, that looks like a correct solution. I didn't work with Alpine, so I didn't know about this feature. But to me this one looks like a cleaner solution:

```js
 <aside x-cloak class="fixed top-0 left-0 z-40 w-64 h-screen transition-transform lg:translate-x-0"
         :class="open ? 'translate-x-0' : '-translate-x-full'"
         aria-label="Sidebar">
 ```
